common-custom-user-data-gradle-plugin/src/main/java/com/gradle/CustomGradleEnterpriseConfig.java

```diff
@@ -50,9 +76,45 @@ static void configureBuildCache(BuildCacheConfiguration buildCache) {
         });
 
         */
+
+        buildCache.local(local -> {
+            withBooleanSystemProp(LOCAL_CACHE_ENABLED_PROP, local::setEnabled);
+            withSystemProp(LOCAL_CACHE_DIRECTORY_PROP, local::setDirectory);
+            withSystemProp(LOCAL_CACHE_CLEANUP_RETENTION_PROP, value -> {
+                Duration retention = Duration.parse(System.getProperty(LOCAL_CACHE_CLEANUP_RETENTION_PROP));
+                local.setRemoveUnusedEntriesAfterDays((int) retention.toDays());
+            });
+            withBooleanSystemProp(LOCAL_CACHE_CLEANUP_ENABLED_PROP, localCacheCleanupEnabled -> {
+                if(!localCacheCleanupEnabled) {
+                    local.setRemoveUnusedEntriesAfterDays(Integer.MAX_VALUE);
+                }
+            });
+        });
+
+        withSystemProp(REMOTE_CACHE_URL_PROP, value -> {
+            buildCache.remote(HttpBuildCache.class).setUrl(value);
+        });
+        withBooleanSystemProp(REMOTE_CACHE_ENABLED_PROP, value -> {
+            buildCache.remote(HttpBuildCache.class).setEnabled(value);
+        });
+        withBooleanSystemProp(REMOTE_CACHE_PUSH_ENABLED_PROP, value -> {
+            buildCache.remote(HttpBuildCache.class).setPush(value);
+        });
+        withBooleanSystemProp(REMOTE_CACHE_ALLOW_UNTRUSTED_SERVER_PROP, value -> {
+            buildCache.remote(HttpBuildCache.class).setAllowUntrustedServer(value);
+        });
     }
 
-    private CustomGradleEnterpriseConfig() {
+    private void withSystemProp(String systemPropertyName, Consumer<String> action) {
+        Provider<String> prop = providers.systemProperty(systemPropertyName).forUseAtConfigurationTime();
```

Comment:
Good feedback!

1. This is done. I left some wrapper functions behind in CustomGradleEnterpriseConfig so that we don't have to pass the ProvidersFactory in every single call to withSystemProp and withBooleanSystemProp.

2. Thankfully, I have a Gradle 5.x project I can use to test this. I tested it, and it indeed does not work. Luckily, I think I can make it work with both Gradle 5.x and the configuration cache.

Reply:
I just pushed a commit that fixes this to work on earlier versions of Gradle as well.

Reply:
Looks great! Thanks.